### PR TITLE
Support etype 17 and 18 in KRB5 AS-REP format

### PR DESCRIPTION
**Pull request by kholia**
I have marked the problematic areas with `XXX`. I am still working on fixing them.

The salt handling part needs some attention. Is there an upper limit on the length of AS-REP messages?
 I am guessing that they are normally much smaller than the TGP-REP messages.

The `split` method is terrible but it seems to work (TM).

Adding similar support to KRB5 TGS format should also be possible.

[AS-REP-sample-catures.zip](https://github.com/magnumripper/JohnTheRipper/files/1394004/AS-REP-sample-catures.zip) <-- sample data for testing.

This PR is partially related to https://github.com/magnumripper/JohnTheRipper/issues/2809 and https://github.com/magnumripper/JohnTheRipper/issues/2721.

[kerberos-sequence-diagram.pdf](https://github.com/magnumripper/JohnTheRipper/files/1394480/kerberos-sequence-diagram.pdf) <- this is really useful.

### CODE REVIEW

**Comment by magnumripper** [comment]
I see no `XXX`... did you get rid of the problems already?

**Comment by kholia** [comment]
I think I fixed all the problems, and forgot to update my message about `XXX` markers.

**Comment by kholia** [comment]
Overall, this AS-REP cracking stuff seems to be much more general-purpose and powerful compared to the TGS-REP stuff.

File: src/krb5_asrep_fmt_plug.c [hunk 1]
```diff
@@ -110,30 +142,58 @@ static size_t cracked_size;
...
 
 static struct custom_salt {
-	dyna_salt dsalt;
-	unsigned char edata1[16];
 	uint32_t edata2len;
-	unsigned char* edata2;
+	unsigned char edata1[16];
+	char salt[256];
+	unsigned char edata2[4096]; // XXX is this large enough?
+	unsigned char etype;
```
Comment (magnumripper) [inline src/krb5_asrep_fmt_plug.c]: This is a bad change. You should keep using dyna_salt - that means there is no "large enough" questions. Salt will be as large as needed.

File: src/krb5_asrep_fmt_plug.c [hunk 2]
```diff
@@ -367,7 +553,7 @@ struct fmt_main fmt_krb5asrep = {
 		SALT_ALIGN,
 		MIN_KEYS_PER_CRYPT,
 		MAX_KEYS_PER_CRYPT,
-		FMT_CASE | FMT_8_BIT | FMT_UNICODE | FMT_UTF8 | FMT_OMP | FMT_DYNA_SALT,
+		FMT_CASE | FMT_8_BIT | FMT_UNICODE | FMT_UTF8 | FMT_OMP,
 		{NULL},
```
Comment (magnumripper) [inline src/krb5_asrep_fmt_plug.c]: Given your "large enough" of 4096 (even when changed back to dyna_salt), I presume FMT_HUGE_INPUT is required as well.

Comment (kholia) [reply src/krb5_asrep_fmt_plug.c]: Yes, this particular change is not good.

However, I was running into the "Could not find salt in db - salt() return inconsistent?" error message with dyna_salt stuff. 

I can post a dyna_salt version of this PR to get some debugging help.

Comment (magnumripper) [reply src/krb5_asrep_fmt_plug.c]: Please do. You shouldn't break things just to work around your own bugs 😉

**Comment by kholia** [comment]
Here is the patch to use dyna_salt,

``` diff
diff --git a/src/krb5_asrep_fmt_plug.c b/src/krb5_asrep_fmt_plug.c
index 4cfc832..e19f34b 100644
--- a/src/krb5_asrep_fmt_plug.c
+++ b/src/krb5_asrep_fmt_plug.c
@@ -76,6 +76,7 @@ john_register_one(&fmt_krb5asrep);
 #include "md4.h"
 #include "hmacmd5.h"
 #include "johnswap.h"
+#include "dyna_salt.h"
 #include "rc4.h"
 #include "hmac_sha.h"
 #include "pbkdf2_hmac_sha1.h"
@@ -99,8 +100,8 @@ john_register_one(&fmt_krb5asrep);
 #define PLAINTEXT_LENGTH        125
 #define BINARY_SIZE             0
 #define BINARY_ALIGN            MEM_ALIGN_NONE
-#define SALT_SIZE               sizeof(struct custom_salt)
-#define SALT_ALIGN              sizeof(uint32_t)
+#define SALT_SIZE               sizeof(struct custom_salt *)
+#define SALT_ALIGN              sizeof(struct custom_salt *)
 #ifdef SIMD_COEF_32
 #define MIN_KEYS_PER_CRYPT      SSE_GROUP_SZ_SHA1
 #define MAX_KEYS_PER_CRYPT      SSE_GROUP_SZ_SHA1
@@ -142,11 +143,12 @@ static size_t cracked_size;
 static int new_keys;
 
 static struct custom_salt {
+	dyna_salt dsalt;
 	uint32_t edata2len;
 	unsigned char edata1[16];
 	char salt[256];
-	unsigned char edata2[4096]; // XXX is this large enough?
 	unsigned char etype;
+	unsigned char *edata2;
 } *cur_salt;
 
 static unsigned char constant[16];
@@ -314,10 +316,12 @@ static void *get_salt(char *ciphertext)
 	char *p;
 	char *ctcopy;
 	char *keeptr;
+	static void *ptr;
 
 	ctcopy = strdup(ciphertext);
 	keeptr = ctcopy;
 	memset(&cs, 0, sizeof(cs));
+	cs.edata2 = NULL;
 
 	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LEN) == 0) {
 		ctcopy += FORMAT_TAG_LEN;
@@ -342,6 +346,7 @@ static void *get_salt(char *ciphertext)
 		// encrypted data
 		p = strtokm(NULL, "$");
 		cs.edata2len = strlen(p) / 2;
+		cs.edata2 = (unsigned char*)mem_calloc_tiny(cs.edata2len + 1, sizeof(char));
 		for (i = 0; i < cs.edata2len; i++) { /* assume edata2 */
 			cs.edata2[i] =
 				atoi16[ARCH_INDEX(p[i * 2])] * 16 +
@@ -366,10 +371,8 @@ static void *get_salt(char *ciphertext)
 		p += strlen(p) + 1;
 
 		/* retrieve non-constant length of edata2 */
-		for (i = 0; p[i] != '\0'; i++)
-			;
-		cs.edata2len = i / 2;
-
+		cs.edata2len = strlen(p) / 2;
+		cs.edata2 = (unsigned char*)mem_calloc_tiny(cs.edata2len + 1, sizeof(char));
 		for (i = 0; i < cs.edata2len; i++) { /* assume edata2 */
 			cs.edata2[i] =
 				atoi16[ARCH_INDEX(p[i * 2])] * 16 +
@@ -379,12 +382,20 @@ static void *get_salt(char *ciphertext)
 
 	MEM_FREE(keeptr);
 
-	return (void *)&cs;
+	/* following is used to fool dyna_salt stuff */
+	cs.dsalt.salt_cmp_offset = SALT_CMP_OFF(struct custom_salt, edata1);
+	cs.dsalt.salt_cmp_size = SALT_CMP_SIZE(struct custom_salt, edata1, edata2len, 0);
+	cs.dsalt.salt_alloc_needs_free = 0;
+
+	ptr = mem_alloc_tiny(sizeof(struct custom_salt), MEM_ALIGN_WORD);
+	memcpy(ptr, &cs, sizeof(struct custom_salt));
+
+	return (void *) &ptr;
 }
 
 static void set_salt(void *salt)
 {
-	cur_salt = (struct custom_salt*)salt;
+	cur_salt = *(struct custom_salt**)salt;
 }
 
 static void set_key(char *key, int index)
@@ -553,7 +564,7 @@ struct fmt_main fmt_krb5asrep = {
 		SALT_ALIGN,
 		MIN_KEYS_PER_CRYPT,
 		MAX_KEYS_PER_CRYPT,
-		FMT_CASE | FMT_8_BIT | FMT_UNICODE | FMT_UTF8 | FMT_OMP,
+		FMT_CASE | FMT_8_BIT | FMT_UNICODE | FMT_UTF8 | FMT_OMP | FMT_DYNA_SALT,
 		{NULL},
 		{ FORMAT_TAG },
 		tests
@@ -571,7 +582,7 @@ struct fmt_main fmt_krb5asrep = {
 		{
 			fmt_default_binary_hash
 		},
-		fmt_default_salt_hash,
+		fmt_default_dyna_salt_hash,
 		NULL,
 		set_salt,
 		set_key,
```

Currently, this patch is crashing with a segmentation fault in `fmt_default_dyna_salt_hash` function. I guess I will have to really learn how to use the dyna_salt stuff.

**Comment by kholia** [comment]
Update: By ordering the members of `struct custom_salt` properly, I will able to get dyna_salt stuff working just fine. Latest commit uses the dyna_salt stuff again. @magnumripper Thanks for the nudge 👍

**Comment by magnumripper** [comment]
I will look into this later, unless you find the bug yourself meanwhile!

**Comment by magnumripper** [comment]
Oh, that's great you already found it! 👍

**Comment by magnumripper** [comment]
I think you still need to add FMT_HUGE_INPUT

**Comment by kholia** [comment]
> I think you still need to add FMT_HUGE_INPUT.

Done now, thanks!

**Pull request merged**